File: ontos/src/frontend/src/types/comments.ts
export enum CommentStatus {
  ACTIVE = "active",
  DELETED = "deleted"
}

export type CommentType = 'comment' | 'rating';

export interface Comment {
  id: string;
  entity_id: string;
  entity_type: string;
  title?: string | null;
  comment: string;
  audience?: string[] | null; // Audience tokens: plain groups, 'team:<team_id>', or 'role:<role_name>'
  project_id?: string | null; // Project ID to scope the comment
  status: CommentStatus;
  comment_type: CommentType;
  rating?: number | null; // 1-5 stars (only for rating type)
  created_by: string;
  updated_by?: string | null;
  created_at: string; // ISO string format
  updated_at: string; // ISO string format
}

export interface CommentCreate {
  entity_id: string;
  entity_type: string;
  title?: string | null;
  comment: string;
  audience?: string[] | null; // Audience tokens: plain groups, 'team:<team_id>', or 'role:<role_name>'
  project_id?: string | null; // Project ID to scope the comment
}

export interface CommentUpdate {
  title?: string | null;
  comment?: string | null;
  audience?: string[] | null;
}

export interface CommentListResponse {
  comments: Comment[];
  total_count: number;
  visible_count: number; // Number of comments visible to current user
}

export interface CommentPermissions {
  can_modify: boolean;
  is_admin: boolean;
}

// Props for comment-related components
export interface CommentSidebarProps {
  entityType: string;
  entityId: string;
  isOpen: boolean;
  onToggle: () => void;
  className?: string;
  fetchCountOnMount?: boolean;
}

export interface CommentItemProps {
  comment: Comment;
  canModify: boolean;
  onEdit: (comment: Comment) => void;
  onDelete: (commentId: string) => void;
}

// Types for audience selection
export interface AudienceTeam {
  id: string;
  name: string;
}

export interface AudienceRole {
  name: string;
}

// Rating-related types
export interface RatingCreate {
  entity_id: string;
  entity_type: string;
  rating: number; // 1-5
  comment?: string | null; // Optional review text
  project_id?: string | null;
}

export interface RatingAggregation {
  entity_type: string;
  entity_id: string;
  average_rating: number;
  total_ratings: number;
  distribution: Record<number, number>; // {1: count, 2: count, ..., 5: count}
  user_current_rating?: number | null; // Current user's latest rating
}

File: ontos/src/backend/src/models/comments.py
from __future__ import annotations
from typing import Optional, List, Dict
from uuid import UUID
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field, model_validator


class CommentStatus(str, Enum):
    ACTIVE = "active"
    DELETED = "deleted"


class CommentType(str, Enum):
    COMMENT = "comment"
    RATING = "rating"


class CommentBase(BaseModel):
    entity_id: str
    entity_type: str = Field(..., description="Type of entity being commented on (data_domain, data_product, data_contract, etc.)")
    title: Optional[str] = Field(None, max_length=255, description="Optional title for the comment")
    comment: str = Field(..., min_length=1, description="The comment content")
    audience: Optional[List[str]] = Field(None, description="List of audience tokens: plain groups, 'team:<team_id>', or 'role:<role_name>'. If null, visible to all users with access to the entity")
    project_id: Optional[str] = Field(None, description="Project ID to scope the comment. If null, visible globally (admin/owning team only)")
    comment_type: CommentType = Field(default=CommentType.COMMENT, description="Type: 'comment' or 'rating'")
    rating: Optional[int] = Field(None, ge=1, le=5, description="Star rating 1-5 (only for rating type)")
    
    @model_validator(mode='after')
    def validate_rating_fields(self):
        """Validate that rating is required for rating-type and forbidden for comment-type."""
        if self.comment_type == CommentType.RATING and self.rating is None:
            raise ValueError('Rating is required for rating-type comments')
        if self.comment_type == CommentType.COMMENT and self.rating is not None:
            raise ValueError('Rating should not be set for regular comments')
        return self


class CommentCreate(CommentBase):
    pass


class CommentUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    comment: Optional[str] = Field(None, min_length=1)
    audience: Optional[List[str]] = None


class Comment(CommentBase):
    id: UUID
    status: CommentStatus = CommentStatus.ACTIVE
    project_id: Optional[str] = None
    comment_type: CommentType = CommentType.COMMENT
    rating: Optional[int] = None
    created_by: str
    updated_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }
    
    @model_validator(mode='after')
    def validate_rating_fields(self):
        """For response models, we allow any combination (DB might have legacy data)."""
        # No strict validation on response - trust the database
        return self


class CommentListResponse(BaseModel):
    """Response model for listing comments with metadata"""
    comments: List[Comment]
    total_count: int
    visible_count: int  # Number of comments visible to current user


class RatingCreate(BaseModel):
    """Simplified model for creating ratings"""
    entity_id: str
    entity_type: str
    rating: int = Field(..., ge=1, le=5, description="Star rating 1-5")
    comment: Optional[str] = Field(None, description="Optional review text")
    project_id: Optional[str] = None


class RatingAggregation(BaseModel):
    """Aggregated rating statistics for an entity"""
    entity_type: str
    entity_id: str
    average_rating: float
    total_ratings: int
    distribution: Dict[int, int]  # {1: count, 2: count, ..., 5: count}
    user_current_rating: Optional[int] = None  # Current user's latest rating